From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Wed, 10 Feb 2021 10:39:38 -0600
Subject: [PATCH] AntiPhysics crash which will be cause by a StackOverflowError


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index c8ed8652eef62fa438fb6f9a14160c67cb04796b..24e3838cfa4ce43f84d4d12d7b240e27bd4e6eba 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -177,6 +177,11 @@ public class DionysusConfig {
         return config.getString(path, config.getString(path));
     }
 
+    public static int maxPhysicsRecursion = 100;
+    private static void maxPhysicsRecursion() {
+        maxPhysicsRecursion = getInt("physics.recursion", 100);
+    }
+
 //    private static void flyingKickMessages() {
 //        flyingKickPlayerMessage = getString("messages.kick.flying-player", flyingKickPlayerMessage);
 //        flyingKickVehicleMessage = getString("messages.kick.flying-vehicle", flyingKickVehicleMessage);
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 922f27b7a5c988613ed5065c3f02c8e52b0127e0..a0f41504cdc930db45435161d90773de31bd912a 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -40,6 +40,10 @@ import com.destroystokyo.paper.antixray.ChunkPacketBlockControllerAntiXray; // A
 import com.google.common.collect.Sets;
 // Paper end
 
+// Dionysus start
+import dev.pomf.dionysus.DionysusConfig;
+// Dionysus end
+
 public abstract class World implements IBlockAccess {
 
     private int a = 63;
@@ -539,7 +543,18 @@ public abstract class World implements IBlockAccess {
         this.b(blockposition.south(), block, blockposition);
     }
 
+    // Dionysus - Prevent StackOverflowError
+    private int recursivePhysicsCount = 0;
+    private static final int MAX_PHYSICS_COUNT = DionysusConfig.maxPhysicsRecursion;
     public void applyPhysics(BlockPosition blockposition, Block block, boolean flag) {
+        // Dionysus Start - Prevent StackOverflowError
+        if (++recursivePhysicsCount >= MAX_PHYSICS_COUNT) {
+            // Dionysus - Reset count once we are done with applyingPhysics
+            recursivePhysicsCount = 0;
+            MinecraftServer.LOGGER.warn("Breaking out of recursive physics loop happening with {} at {}", block.getName(), blockposition.toString());
+            return;
+        }
+        // Dionysus End
         if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
         this.a(blockposition.west(), block, blockposition);
         this.a(blockposition.east(), block, blockposition);
@@ -551,6 +566,9 @@ public abstract class World implements IBlockAccess {
             this.c(blockposition, block);
         }
 
+        // Dionysus - Reset count once we are done with applyingPhysics
+        recursivePhysicsCount = 0;
+
         this.chunkPacketBlockController.updateNearbyBlocks(this, blockposition); // Paper - Anti-Xray
     }
 
