From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Fri, 5 Feb 2021 00:00:10 -0600
Subject: [PATCH] Dionysus config

API to retrieve raw YamlConfiguration + timing exports

diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2912bdae68f52a0c3b2c58e849df5271ef2170bf..fc294a94d9c33bc256fdaee1d3a91eeeeaedc6a6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1015,6 +1015,12 @@ public interface Server extends PluginMessageRecipient {
             throw new UnsupportedOperationException("Not supported yet.");
         }
 
+        // Dionysus start - add config to timings report
+        public org.bukkit.configuration.file.YamlConfiguration getDionysusConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
         /**
          * Sends the component to the player
          *
