From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Fri, 5 Feb 2021 00:18:18 -0600
Subject: [PATCH] Change Math.sin and Math.cos to respective MathHelper


diff --git a/src/main/java/net/minecraft/server/BiomeMesa.java b/src/main/java/net/minecraft/server/BiomeMesa.java
index 67f8ad8eda987217291dccdbc61419ffed89725a..c5472b50109c2efb46c896d097860e48e65356ca 100644
--- a/src/main/java/net/minecraft/server/BiomeMesa.java
+++ b/src/main/java/net/minecraft/server/BiomeMesa.java
@@ -89,7 +89,7 @@ public class BiomeMesa extends BiomeBase {
         IBlockData iblockdata = BiomeMesa.A;
         IBlockData iblockdata1 = this.r;
         int j1 = (int) (d0 / 3.0D + 3.0D + random.nextDouble() * 0.25D);
-        boolean flag = Math.cos(d0 / 3.0D * 3.141592653589793D) > 0.0D;
+        boolean flag = MathHelper.cos((float)d0 / 3.0f * 3.141592653589793f) > 0.0f;
         int k1 = -1;
         boolean flag1 = false;
         int l1 = 0;
diff --git a/src/main/java/net/minecraft/server/EntityEnderSignal.java b/src/main/java/net/minecraft/server/EntityEnderSignal.java
index 6a5e20ab11923ef92af941afd69467224f473e64..5d7b85759db9cc439968dadaf46520fb1c6a706f 100644
--- a/src/main/java/net/minecraft/server/EntityEnderSignal.java
+++ b/src/main/java/net/minecraft/server/EntityEnderSignal.java
@@ -85,8 +85,8 @@ public class EntityEnderSignal extends Entity {
                 this.motY *= 0.8D;
             }
 
-            this.motX = Math.cos((double) f2) * d2;
-            this.motZ = Math.sin((double) f2) * d2;
+            this.motX = MathHelper.cos(f2) * d2;
+            this.motZ = MathHelper.sin(f2) * d2;
             if (this.locY < this.b) {
                 this.motY += (1.0D - this.motY) * 0.014999999664723873D;
             } else {
diff --git a/src/main/java/net/minecraft/server/EntityMinecartAbstract.java b/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
index beefd53d0e0e8b742af6386251427872f10ccf05..761f34c7b62468f0b2f51df3151e53f27dfd2931 100644
--- a/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
+++ b/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
@@ -487,8 +487,8 @@ public abstract class EntityMinecartAbstract extends Entity implements INamableT
         if (entity instanceof EntityLiving) {
             d6 = ((EntityLiving) entity).bg;
             if (d6 > 0.0D) {
-                d7 = -Math.sin((double) (entity.yaw * 0.017453292F));
-                d8 = Math.cos((double) (entity.yaw * 0.017453292F));
+                d7 = -MathHelper.sin(entity.yaw * 0.017453292F);
+                d8 = MathHelper.cos(entity.yaw * 0.017453292F);
                 d9 = this.motX * this.motX + this.motZ * this.motZ;
                 if (d9 < 0.01D) {
                     this.motX += d7 * 0.1D;
diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index 5eabbd2063b09fd1822a22d1e326b72f581d8ca9..b1320ec746b51f12aeaa5adb6191f4bf42ca2d61 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -25,9 +25,9 @@ public class EntityTNTPrimed extends Entity {
         this.setPosition(d0, d1, d2);
         float f = (float) (Math.random() * 6.2831854820251465D);
 
-        this.motX = (double) (-((float) Math.sin((double) f)) * 0.02F);
+        this.motX = (double) (-MathHelper.sin(f) * 0.02F);
         this.motY = 0.20000000298023224D;
-        this.motZ = (double) (-((float) Math.cos((double) f)) * 0.02F);
+        this.motZ = (double) (-MathHelper.cos(f) * 0.02F);
         this.setFuseTicks(80);
         this.lastX = d0;
         this.lastY = d1;
diff --git a/src/main/java/net/minecraft/server/WorldGenStronghold.java b/src/main/java/net/minecraft/server/WorldGenStronghold.java
index d6f8396c327ffbf5865001549d2325410459c1fd..f47046889d3073cbd55cc0807b8b4e9b66a12ffd 100644
--- a/src/main/java/net/minecraft/server/WorldGenStronghold.java
+++ b/src/main/java/net/minecraft/server/WorldGenStronghold.java
@@ -130,8 +130,8 @@ public class WorldGenStronghold extends StructureGenerator {
         if (l < this.d.length) {
             for (int i1 = 0; i1 < this.d.length; ++i1) {
                 double d1 = 4.0D * this.h + this.h * (double) j * 6.0D + (random.nextDouble() - 0.5D) * this.h * 2.5D;
-                int j1 = (int) Math.round(Math.cos(d0) * d1);
-                int k1 = (int) Math.round(Math.sin(d0) * d1);
+                int j1 = (int) Math.round(MathHelper.cos((float)d0) * d1);
+                int k1 = (int) Math.round(MathHelper.sin((float)d0) * d1);
                 BlockPosition blockposition = this.g.getWorldChunkManager().a((j1 << 4) + 8, (k1 << 4) + 8, 112, this.a, random);
 
                 if (blockposition != null) {
diff --git a/src/main/java/net/minecraft/server/WorldProvider.java b/src/main/java/net/minecraft/server/WorldProvider.java
index a0ef408148383405cbcdb9f3c6372ef5bcebc984..1ffd02e8ed41cf3cde91fdb0bb3a0c151f0c5851 100644
--- a/src/main/java/net/minecraft/server/WorldProvider.java
+++ b/src/main/java/net/minecraft/server/WorldProvider.java
@@ -76,7 +76,7 @@ public abstract class WorldProvider {
 
         float f2 = f1;
 
-        f1 = 1.0F - (float) ((Math.cos((double) f1 * 3.141592653589793D) + 1.0D) / 2.0D);
+        f1 = 1.0F - (MathHelper.cos(f1 * 3.141592653589793f) + 1.0f) / 2.0f;
         f1 = f2 + (f1 - f2) / 3.0F;
         return f1;
     }
