From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Mon, 8 Feb 2021 00:02:46 -0600
Subject: [PATCH] AntiBackdoor


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index c8ed8652eef62fa438fb6f9a14160c67cb04796b..6f03a641bcc66bb36ec8396212207a96c0f95904 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -28,8 +28,7 @@ public class DionysusConfig {
     private static final Pattern NOT_NUMERIC = Pattern.compile("[^-\\d.]");
     /*========================================================================*/
     public static YamlConfiguration config;
-    public static String flyingKickPlayerMessage = "Flying is not enabled on this server";
-    public static String flyingKickVehicleMessage = "Flying is not enabled on this server";
+//    public static String backdoorKickPlayerMessage = "Flying is not enabled on this server";
     /*========================================================================*/
     static int version;
     static Map<String, Command> commands;
@@ -167,6 +166,7 @@ public class DionysusConfig {
         return config.getInt(path, config.getInt(path));
     }
 
+    @SuppressWarnings("unchecked")
     private static <T> List<T> getList(String path, T def) {
         config.addDefault(path, def);
         return (List<T>) config.getList(path, config.getList(path));
@@ -177,8 +177,14 @@ public class DionysusConfig {
         return config.getString(path, config.getString(path));
     }
 
-//    private static void flyingKickMessages() {
-//        flyingKickPlayerMessage = getString("messages.kick.flying-player", flyingKickPlayerMessage);
-//        flyingKickVehicleMessage = getString("messages.kick.flying-vehicle", flyingKickVehicleMessage);
+    public static boolean enableAntiBackdoor = true;
+
+    private static void antiBackdoor() {
+        enableAntiBackdoor = getBoolean("anti-backdoor", true);
+    }
+
+    // NOTE: leaving this here just in case we want to change this to a kick or notification
+//    private static void backdoorMessages() {
+//        backdoorKickPlayerMessage = getString("messages.kick.backdoor", backdoorKickPlayerMessage);
 //    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index e49fa6e4f228555e276128d85a1cac665dc42e7b..eaeae5ca831d9a8250c9d60bc24d10caad16c7f2 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -28,6 +28,8 @@ import org.bukkit.event.player.PlayerGameModeChangeEvent;
 import org.bukkit.event.player.PlayerLocaleChangeEvent;
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.inventory.MainHand;
+
+import static dev.pomf.dionysus.DionysusConfig.enableAntiBackdoor;
 // CraftBukkit end
 
 public class EntityPlayer extends EntityHuman implements ICrafting {
@@ -1219,6 +1221,16 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
 
     public void a(EnumGamemode enumgamemode) {
+        // Dionysus start - Fix "backdoor" exploit
+        // NOTE: This does not work on servers where everyone is in Adventure Mode
+        if (enableAntiBackdoor
+                && (enumgamemode != EnumGamemode.SURVIVAL)) {
+            // Force send update to client
+            this.playerInteractManager.setGameMode(EnumGamemode.SURVIVAL);
+            return;
+        }
+        // Dionysus end
+
         // CraftBukkit start
         if (enumgamemode == this.playerInteractManager.getGameMode()) {
             return;
