From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Thu, 11 Feb 2021 01:02:59 -0600
Subject: [PATCH] Customizable Deflate Compression


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 0e542ad19d2e3d8c2762091d22f32f62056ddd1c..48464f4d1b3da414d3f9b9d86c0e53c08e774a80 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -195,4 +195,13 @@ public class DionysusConfig {
         animalAiLimitTPS = getDouble("ai-limit.animal.tps", 6);
         animalAiLimitPlayerCount = getInt("ai-limit.animal.player-count", 120);
     }
+
+    public static int compressionLevel = -1;
+    private static void compressionLevel() {
+        compressionLevel = getInt("compression-level", -1);
+        if (compressionLevel < -1 || compressionLevel > 9) {
+            logError("Compression level is invalid resetting to default. Valid compression levels are -1 (Default), 0 (None), 1 (Fast) through 9 (Best)");
+            compressionLevel = -1;
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java b/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java
index 8b5aeb1b02eeb24b29fede44b625587826052f90..76858edadf56a014bb84228689931f66fcaeee26 100644
--- a/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java
+++ b/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import dev.pomf.dionysus.DionysusConfig;
+
 import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
 import java.io.DataInput;
@@ -29,7 +31,7 @@ public class NBTCompressedStreamTools {
     }
 
     public static void a(NBTTagCompound nbttagcompound, OutputStream outputstream) throws IOException {
-        DataOutputStream dataoutputstream = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(outputstream)));
+        DataOutputStream dataoutputstream = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(outputstream) {{ def.setLevel(DionysusConfig.compressionLevel);}})); // Dionysus - NBT Compression Level
 
         try {
             a(nbttagcompound, (DataOutput) dataoutputstream);
diff --git a/src/main/java/net/minecraft/server/PacketCompressor.java b/src/main/java/net/minecraft/server/PacketCompressor.java
index dc87fe47fa7f91643630f9e257e4facb1c0e592a..a80d293b8a37ff86a243d6167a30639f7105749c 100644
--- a/src/main/java/net/minecraft/server/PacketCompressor.java
+++ b/src/main/java/net/minecraft/server/PacketCompressor.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import dev.pomf.dionysus.DionysusConfig;
 import io.netty.buffer.ByteBuf;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.MessageToByteEncoder;
@@ -14,19 +15,20 @@ public class PacketCompressor extends MessageToByteEncoder<ByteBuf> {
     public PacketCompressor(int i) {
         this.c = i;
         this.b = new Deflater();
+        this.b.setLevel(DionysusConfig.compressionLevel); // Dionysus - Packet Compression Level
     }
 
-    protected void a(ChannelHandlerContext channelhandlercontext, ByteBuf bytebuf, ByteBuf bytebuf1) throws Exception {
-        int i = bytebuf.readableBytes();
-        PacketDataSerializer packetdataserializer = new PacketDataSerializer(bytebuf1);
+    protected void a(ChannelHandlerContext channelhandlercontext, ByteBuf incoming, ByteBuf outgoing) throws Exception {
+        int i = incoming.readableBytes();
+        PacketDataSerializer packetdataserializer = new PacketDataSerializer(outgoing);
 
         if (i < this.c) {
             packetdataserializer.d(0);
-            packetdataserializer.writeBytes(bytebuf);
+            packetdataserializer.writeBytes(incoming);
         } else {
             byte[] abyte = new byte[i];
 
-            bytebuf.readBytes(abyte);
+            incoming.readBytes(abyte);
             packetdataserializer.d(abyte.length);
             this.b.setInput(abyte, 0, i);
             this.b.finish();
@@ -47,7 +49,7 @@ public class PacketCompressor extends MessageToByteEncoder<ByteBuf> {
     }
 
     @Override
-    protected void encode(ChannelHandlerContext channelhandlercontext, ByteBuf object, ByteBuf bytebuf) throws Exception {
-        this.a(channelhandlercontext, object, bytebuf);
+    protected void encode(ChannelHandlerContext channelhandlercontext, ByteBuf incoming, ByteBuf outgoing) throws Exception {
+        this.a(channelhandlercontext, incoming, outgoing);
     }
 }
