From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Sun, 7 Mar 2021 17:44:35 -0600
Subject: [PATCH] Fair Natural Spawning


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index bde1d8e5f003bc9aef9131a5f22f624e42955e12..e3ca26e90e956cca655f3449e3985079cc842f46 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -214,4 +214,12 @@ public class DionysusConfig {
     private static void serverModName() {
         serverModName = getString("settings.server-mod-name", serverModName);
     }
+
+    public static boolean fairNaturalSpawns = true;
+    private static void fairNaturalSpawns() {
+        fairNaturalSpawns = getBoolean("fair-natural-spawns", fairNaturalSpawns);
+        if (fairNaturalSpawns) {
+            Bukkit.getLogger().info("Fair natural spawns enabled, using the Spigot world mob-spawn-range for search and Bukkit spawn-limits for the count");
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
index 3b001ae0ae2f15486285a19b3b301c77eefb57e1..36b5d4343eed368003c674d635eaa3681b459914 100644
--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
+++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
@@ -6,6 +6,7 @@ import java.util.Random;
 
 // CraftBukkit start
 import com.destroystokyo.paper.exception.ServerInternalException;
+import dev.pomf.dionysus.DionysusConfig;
 import it.unimi.dsi.fastutil.longs.LongArraySet;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import org.bukkit.craftbukkit.util.LongHash;
@@ -71,6 +72,25 @@ public final class SpawnerCreature {
                     byte b0 = worldserver.spigotConfig.mobSpawnRange;
                     b0 = ( b0 > worldserver.spigotConfig.viewDistance ) ? (byte) worldserver.spigotConfig.viewDistance : b0;
                     b0 = ( b0 > 8 ) ? 8 : b0;
+
+                    // Dionysus start - Limit spawning based on nearby mobs
+                    // NOTE: this incorrectly assumes all player affecting spawning will be monsters.
+                    if (DionysusConfig.fairNaturalSpawns) {
+                        int searchRange = worldserver.spigotConfig.mobSpawnRange << 4;
+                        AxisAlignedBB bb = new AxisAlignedBB(entityhuman.getX() - searchRange, entityhuman.getY() - searchRange, entityhuman.getZ() - searchRange,
+                                entityhuman.getX() + searchRange, entityhuman.getY() + searchRange, entityhuman.getZ() + searchRange);
+
+                        int nearbyCount = worldserver
+                                // this should cover anything that is considered a monster
+                                // EntityFlying doesn't implement IMonster, but EntityGhast does
+                                .getEntities(entityhuman, bb, entity -> entity instanceof IMonster)
+                                .size();
+                        if (nearbyCount > worldserver.getWorld().getMonsterSpawnLimit()) {
+                            continue;
+                        }
+                    }
+                    // Dionysus End
+
                     // Paper start
                     com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent event;
                     event = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(
