From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RemainingToast <remainingtoast@gmail.com>
Date: Fri, 7 May 2021 15:23:50 +1000
Subject: [PATCH] Change Bed Mechanics


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 81f24b09bfc127226918aea6215a79022307d8af..e172491020e7a734e2d43007e9feb816628e50be 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -230,4 +230,13 @@ public class DionysusConfig {
             Bukkit.getLogger().info("Redact player data enabled, IP Address and location data will be removed.");
         }
     }
+
+    public static boolean alternativeBedMechanics = true;
+    private static void alternativeBedMechanics() {
+        alternativeBedMechanics = getBoolean("alternative-bed-mechanics", alternativeBedMechanics);
+        if (alternativeBedMechanics) {
+            Bukkit.getLogger().info("Alternative Bed Mechanics actived. Respawn is set when bed is clicked on.");
+        }
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/BlockBed.java b/src/main/java/net/minecraft/server/BlockBed.java
index 4ea34b1a23202737c3a2527ae97ecd47bf176c43..8363fc0feefd66ac480a4ff3da3736c39edb7266 100644
--- a/src/main/java/net/minecraft/server/BlockBed.java
+++ b/src/main/java/net/minecraft/server/BlockBed.java
@@ -60,6 +60,10 @@ public class BlockBed extends BlockFacingHorizontal implements ITileEntity {
                 if (entityhuman_enumbedresult == EntityHuman.EnumBedResult.OK) {
                     iblockdata = iblockdata.set(BlockBed.OCCUPIED, Boolean.valueOf(true));
                     world.setTypeAndData(blockposition, iblockdata, 4);
+                    // Dionysus - alternative bed mechanics
+                    if (dev.pomf.dionysus.DionysusConfig.alternativeBedMechanics) {
+                        entityhuman.a(new ChatComponentText("Respawn point set"), true);
+                    }
                     return true;
                 } else {
                     if (entityhuman_enumbedresult == EntityHuman.EnumBedResult.NOT_POSSIBLE_NOW) {
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index e06d65a32b3755dddafae1a2cc136ec79d3d6abe..e98077c8caf0fe4e31bf9fbaea1f3a1a3aeb0e76 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1233,7 +1233,8 @@ public abstract class EntityHuman extends EntityLiving {
                 return EntityHuman.EnumBedResult.NOT_POSSIBLE_HERE;
             }
 
-            if (this.world.D()) {
+            // Dionysus - alternative bed mechanics
+            if (this.world.D() && !dev.pomf.dionysus.DionysusConfig.alternativeBedMechanics) {
                 return EntityHuman.EnumBedResult.NOT_POSSIBLE_NOW;
             }
 
