From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RemainingToast <59129680+RemainingToast@users.noreply.github.com>
Date: Wed, 15 Dec 2021 19:39:21 +1000
Subject: [PATCH] Change Bed Mechanics


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 81f24b09bfc127226918aea6215a79022307d8af..e172491020e7a734e2d43007e9feb816628e50be 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -230,4 +230,13 @@ public class DionysusConfig {
             Bukkit.getLogger().info("Redact player data enabled, IP Address and location data will be removed.");
         }
     }
+
+    public static boolean alternativeBedMechanics = true;
+    private static void alternativeBedMechanics() {
+        alternativeBedMechanics = getBoolean("alternative-bed-mechanics", alternativeBedMechanics);
+        if (alternativeBedMechanics) {
+            Bukkit.getLogger().info("Alternative Bed Mechanics actived. Respawn is set when bed is clicked on.");
+        }
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/BlockBed.java b/src/main/java/net/minecraft/server/BlockBed.java
index f1a1079918ae37672b2c5978b82f6392801859d0..d01ab2998eda93808e2dc6e492d2ab86a10e628f 100644
--- a/src/main/java/net/minecraft/server/BlockBed.java
+++ b/src/main/java/net/minecraft/server/BlockBed.java
@@ -60,6 +60,10 @@ public class BlockBed extends BlockFacingHorizontal implements ITileEntity {
                 if (entityhuman_enumbedresult == EntityHuman.EnumBedResult.OK) {
                     iblockdata = iblockdata.set(BlockBed.OCCUPIED, Boolean.valueOf(true));
                     world.setTypeAndData(blockposition, iblockdata, 4);
+                    // Dionysus - alternative bed mechanics
+                    if (dev.pomf.dionysus.DionysusConfig.alternativeBedMechanics) {
+                        entityhuman.a(new ChatComponentText("Respawn point set"), true);
+                    }
                     return true;
                 } else {
                     if (entityhuman_enumbedresult == EntityHuman.EnumBedResult.NOT_POSSIBLE_NOW) {
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index aa0b27f0e1c6cd5168af02157ce8ce495afc6373..6c30d30ff2d7c46aa9e3e32473efc94cb243defa 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1233,7 +1233,8 @@ public abstract class EntityHuman extends EntityLiving {
                 return EntityHuman.EnumBedResult.NOT_POSSIBLE_HERE;
             }
 
-            if (this.world.D()) {
+            // Dionysus - alternative bed mechanics
+            if (this.world.D() && !dev.pomf.dionysus.DionysusConfig.alternativeBedMechanics) {
                 return EntityHuman.EnumBedResult.NOT_POSSIBLE_NOW;
             }
 
