From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RemainingToast <59129680+RemainingToast@users.noreply.github.com>
Date: Wed, 15 Dec 2021 19:39:21 +1000
Subject: [PATCH] Chunk overload dupe config settings


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 0fe33eb470ad0be018d127194273b39cb42be0ff..e0832d7f3a5e8fc992eafc824642b518c1e2810f 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -251,4 +251,13 @@ public class DionysusConfig {
             Bukkit.getLogger().info("Tickless Block Placement Mechanics activated. Block placement based on time instead of ticks.");
         }
     }
+
+    public static boolean chunkOverloadDupe = false;
+    public static boolean chunkOverloadDupeLog = false;
+    private static void chunkOverloadDupe() {
+        chunkOverloadDupe = getBoolean("chunk-overload-dupe.allow", chunkOverloadDupe);
+        if (chunkOverloadDupe) Bukkit.getLogger().info("Chunk overload dupe activated.");
+        chunkOverloadDupeLog = getBoolean("chunk-overload-dupe.log", chunkOverloadDupeLog);
+        if (chunkOverloadDupeLog) Bukkit.getLogger().info("Chunk overload dupes will be logged.");
+    }
 }
diff --git a/src/main/java/net/minecraft/server/RegionFile.java b/src/main/java/net/minecraft/server/RegionFile.java
index 542a35d13014f8540f1bf525f021188b8955fc56..5fecde675808a5c179ff2b055c3faa127b9f962e 100644
--- a/src/main/java/net/minecraft/server/RegionFile.java
+++ b/src/main/java/net/minecraft/server/RegionFile.java
@@ -20,6 +20,8 @@ import java.util.zip.GZIPInputStream;
 import java.util.zip.InflaterInputStream;
 import javax.annotation.Nullable;
 
+import dev.pomf.dionysus.DionysusConfig; // Dionysus
+
 public class RegionFile {
 
     // Spigot start
@@ -198,6 +200,14 @@ public class RegionFile {
             int k1 = (k + 5) / 4096 + 1;
 
             if (k1 >= 256) {
+                // Dionysus start - chunk overload dupe
+                if (DionysusConfig.chunkOverloadDupe) {
+                    if (DionysusConfig.chunkOverloadDupeLog) {
+                        org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.INFO, "Chunk dupe detected: ({0}, {1}) {2}", new Object[]{i, j, this.b});
+                    }
+                    return;
+                }
+                // Dionysus end
                 // Spigot start
                 if (!USE_SPIGOT_OVERSIZED_METHOD && !RegionFileCache.isOverzealous()) throw new ChunkTooLargeException(i, j, k1); // Paper - throw error instead
                 org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING,"Large Chunk Detected: ({0}, {1}) Size: {2} {3}", new Object[]{i, j, k1, this.b});
