From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlexProgrammerDE <40795980+AlexProgrammerDE@users.noreply.github.com>
Date: Thu, 23 Jun 2022 20:24:27 +0200
Subject: [PATCH] Add book nbt max size patch


diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 4d1a8f35a1170a7d9ed522189345c79db457f33b..63b2ffe197bc774a5ab870abe973fc3226986203 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -408,4 +408,10 @@ public class DionysusConfig {
 
         RegionFile.deflater.setLevel(chunkCompressionLevel);
     }
+
+    public static int maxProcessedBookNBTSize = 8000;
+
+    private static void maxProcessedBookNBTSize() {
+        maxProcessedBookNBTSize = getInt("max-processed-book-nbt-size", 8000);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/PacketDataSerializer.java b/src/main/java/net/minecraft/server/PacketDataSerializer.java
index c1273e988e6a46c20eca54530f928404c03ad947..54b518dcd422e842af128aeacee8670d1ea3d7ee 100644
--- a/src/main/java/net/minecraft/server/PacketDataSerializer.java
+++ b/src/main/java/net/minecraft/server/PacketDataSerializer.java
@@ -269,6 +269,11 @@ public class PacketDataSerializer extends ByteBuf {
             ItemStack itemstack = new ItemStack(Item.getById(short0), b0, short1);
 
             itemstack.setTag(this.j());
+            // Dionysus start
+            if (itemstack.getTag() != null && itemstack.getTag().guessSize() > dev.pomf.dionysus.DionysusConfig.maxProcessedBookNBTSize) {
+                return ItemStack.a;
+            }
+            // Dionysus end
             // CraftBukkit start
             if (itemstack.getTag() != null) {
                 CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
