From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: comendantmc <2b2t.org.ru@gmail.com>
Date: Wed, 8 Dec 2021 22:12:27 +0100
Subject: [PATCH] Make bedrock non buildable, fix getBrightness crash


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 6552c1ec04ff9a19548a9fe7fb7f18c25e2d01a8..61bc886cb946a2e6b9e3ba8305b49bebfcc49478 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -618,7 +618,13 @@ public class Chunk {
         int i = blockposition.getX() & 15;
         int j = blockposition.getY();
         int k = blockposition.getZ() & 15;
-        ChunkSection chunksection = this.sections[j >> 4];
+        ChunkSection chunksection;
+        try {
+            chunksection = this.sections[j >> 4];
+        } catch(Exception ex) {
+            System.out.println(ex.getMessage());
+            chunksection = this.sections[(j & 15) >> 4];
+        }
 
         return chunksection == Chunk.a ? (this.c(blockposition) ? enumskyblock.c : 0) : (enumskyblock == EnumSkyBlock.SKY ? (!this.world.worldProvider.m() ? 0 : chunksection.b(i, j & 15, k)) : (enumskyblock == EnumSkyBlock.BLOCK ? chunksection.c(i, j & 15, k) : enumskyblock.c));
     }
diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 42e4b2d01d1f4cb7729580daac0dc7ac34adc382..bbd173fde2ee2c463d976a3781bc8027ea68faa6 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -344,7 +344,8 @@ public class PortalTravelAgent {
                                         int l4 = j2 + (l3 - 1) * j3 - k3 * i3;
 
                                         blockposition_mutableblockposition.c(j4, k4, l4);
-                                        if (i4 < 0 && !this.world.getType(blockposition_mutableblockposition).getMaterial().isBuildable() || i4 >= 0 && !this.world.isEmpty(blockposition_mutableblockposition)) {
+                                        IBlockData iblockdata = this.world.getType(blockposition_mutableblockposition);
+                                        if (i4 < 0 && !(iblockdata.getMaterial().isBuildable() && iblockdata.getBlock() != Blocks.BEDROCK) || i4 >= 0 && !this.world.isEmpty(blockposition_mutableblockposition)) {
                                             continue label271;
                                         }
                                     }
@@ -390,7 +391,8 @@ public class PortalTravelAgent {
                                         j4 = k2 + l3;
                                         k4 = j2 + (k3 - 1) * j3;
                                         blockposition_mutableblockposition.c(i4, j4, k4);
-                                        if (l3 < 0 && !this.world.getType(blockposition_mutableblockposition).getMaterial().isBuildable() || l3 >= 0 && !this.world.isEmpty(blockposition_mutableblockposition)) {
+                                        IBlockData iblockdata = this.world.getType(blockposition_mutableblockposition);
+                                        if (l3 < 0 && !(iblockdata.getMaterial().isBuildable() && iblockdata.getBlock() != Blocks.BEDROCK) || l3 >= 0 && !this.world.isEmpty(blockposition_mutableblockposition)) {
                                             continue label219;
                                         }
                                     }
