From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Sun, 10 Jul 2022 20:46:55 +0500
Subject: [PATCH] Skip-updating-entity-tracker-without-Players


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 14cde8b1ac2fe991f4d721b4b178830eaec53d6f..58e3e3d067798e3dcf6a16e7d1c09cf8edbd8670 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -994,7 +994,9 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
                     }
                 }
                 try { // Dionysus end - Flush consolidation
-                worldserver.getTracker().updatePlayers();
+                if (this.getPlayerList().getPlayerCount() != 0) { // NeonPaper
+                    worldserver.getTracker().updatePlayers();
+                }
                 // Dionysus start - Flush consolidation
                 } finally {
                     for (net.minecraft.server.PlayerConnection playerConnection : disabledFlushes) {
