From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Sun, 10 Jul 2022 20:46:55 +0500
Subject: [PATCH] Skip-updating-entity-tracker-without-Players


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 9bba76909494839d587cdd898830202695bf6985..896e0c7b09de8ec26760730ee23ef7ce8a203295 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -958,7 +958,9 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
                     }
                 }
                 try { // Dionysus end - Flush consolidation
-                worldserver.getTracker().updatePlayers();
+                if (this.getPlayerList().getPlayerCount() != 0) { // NeonPaper
+                    worldserver.getTracker().updatePlayers();
+                }
                 // Dionysus start - Flush consolidation
                 } finally {
                     for (net.minecraft.server.PlayerConnection playerConnection : disabledFlushes) {
