From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Sat, 9 Jul 2022 03:40:49 +0500
Subject: [PATCH] Fix-Teleport-Invisibility


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 52ddc347a645c8293bee39348303ad05dc5f7793..f6b838cb5a55e118cccdc7443699e66de9ee54ab 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -179,7 +179,7 @@ public class EntityTrackerEntry {
                         this.v = 0;
                         // CraftBukkit start - Refresh list of who can see a player before sending teleport packet
                         if (this.tracker instanceof EntityPlayer) {
-                            this.scanPlayers(new java.util.ArrayList(this.trackedPlayers));
+                            this.scanPlayers(new java.util.ArrayList(this.tracker.world.players)); // [NeonPaper] Fix Teleport Invisibility this.trackedPlayers -> this.tracker.world.players
                         }
                         // CraftBukkit end
                         this.c();
