From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 7 Jul 2022 01:55:42 +0500
Subject: [PATCH] Do-not-load-chunks-for-light-checks


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 039b6e9861295f5f2a171b38884e0717dd3871e0..cbc41847c1f3efb21c8a6935593d27f007a8c8e8 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -693,7 +693,7 @@ public abstract class World implements IBlockAccess {
             if (blockposition.getY() >= 256) {
                 blockposition = new BlockPosition(blockposition.getX(), 255, blockposition.getZ());
             }
-
+			if (!this.isLoaded(blockposition)) return 0; // Paper
             return this.getChunkAtWorldCoords(blockposition).a(blockposition, 0);
         }
     }
