From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Fri, 15 Jul 2022 02:55:16 +0500
Subject: [PATCH] fix-some-rails-connecting-improperly


diff --git a/src/main/java/net/minecraft/server/BlockMinecartDetector.java b/src/main/java/net/minecraft/server/BlockMinecartDetector.java
index 872543df7a80e84b38ebdac09c40da8d52be71f4..ebabf820d3ce7cc5969f537a9af1a1e5d98c59ae 100644
--- a/src/main/java/net/minecraft/server/BlockMinecartDetector.java
+++ b/src/main/java/net/minecraft/server/BlockMinecartDetector.java
@@ -60,6 +60,7 @@ public class BlockMinecartDetector extends BlockMinecartTrackAbstract {
     }
 
     private void e(World world, BlockPosition blockposition, IBlockData iblockdata) {
+        if (iblockdata.getBlock() != this) { return; } // Dionysus - not our block, don't do anything
         boolean flag = ((Boolean) iblockdata.get(BlockMinecartDetector.POWERED)).booleanValue();
         boolean flag1 = false;
         List list = this.a(world, blockposition, EntityMinecartAbstract.class, new Predicate[0]);
