From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 26 Apr 2021 09:50:16 +0200
Subject: [PATCH] (KigPaper) Kick after receiving invalid creative action
 packet


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index bde00b896d2d8e922068346e6cba7ed2f5dd7494..9f8f33d6f99afb8f93040375a4bac9f21e122551 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2222,6 +2222,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     entityitem.j();
                 }
             }
+        } else {
+            disconnect("Not in creative mode"); // KigPaper
         }
 
     }
