From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Mon, 11 Jul 2022 19:51:44 +0500
Subject: [PATCH] Better-handling-of-some-ItemMeta


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
index 7b3567894468d7c38c364c6f2be7d8d7c78e860a..cada80c314894ab7bded6db1e64a8ca5c47ffeb6 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
@@ -53,7 +53,14 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
             NBTTagList patterns = entityTag.getList(PATTERNS.NBT, CraftMagicNumbers.NBT.TAG_COMPOUND);
             for (int i = 0; i < Math.min(patterns.size(), 20); i++) {
                 NBTTagCompound p = patterns.get(i);
-                this.patterns.add(new Pattern(DyeColor.getByDyeData((byte) p.getInt(COLOR.NBT)), PatternType.getByIdentifier(p.getString(PATTERN.NBT))));
+                // NeonPaper start - Backport SPIGOT-5428
+                DyeColor color = DyeColor.getByDyeData((byte) p.getInt(COLOR.NBT));
+                PatternType pattern = PatternType.getByIdentifier(p.getString(PATTERN.NBT));
+
+                if (color != null && pattern != null) {
+                    this.patterns.add(new Pattern(color, pattern));
+                }
+                // NeonPaper end
             }
         }
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java
index de695e4e7bef8167037572fc971476adeade33cc..01f9bc75857ce69048040501adcd33bd8ccc5a2d 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java
@@ -55,7 +55,13 @@ class CraftMetaMap extends CraftMetaItem implements MapMeta {
             }
 
             if (display.hasKey(MAP_COLOR.NBT)) {
-                color = Color.fromRGB(display.getInt(MAP_COLOR.NBT));
+                // NeonPaper start - Backport SPIGOT-5428
+                try {
+                    color = Color.fromRGB(display.getInt(MAP_COLOR.NBT));
+                } catch (IllegalArgumentException ex) {
+                    // Invalid colour
+                }
+                // NeonPaper end
             }
         }
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java
index 1080e52ccaf4e6dd50e5bf0034f24f879e119673..6e801ab280ba432793e53ac0b838e0d2f0e29874 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java
@@ -61,7 +61,13 @@ class CraftMetaPotion extends CraftMetaItem implements PotionMeta {
             type = CraftPotionUtil.toBukkit(tag.getString(DEFAULT_POTION.NBT));
         }
         if (tag.hasKey(POTION_COLOR.NBT)) {
-            color = Color.fromRGB(tag.getInt(POTION_COLOR.NBT));
+            // NeonPaper start - Backport SPIGOT-5428
+            try {
+                color = Color.fromRGB(tag.getInt(POTION_COLOR.NBT));
+            } catch (IllegalArgumentException ex) {
+                // Invalid colour
+            }
+            // NeonPaper end
         }
         if (tag.hasKey(POTION_EFFECTS.NBT)) {
             NBTTagList list = tag.getList(POTION_EFFECTS.NBT, CraftMagicNumbers.NBT.TAG_COMPOUND);
