From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Mon, 18 Jul 2022 12:40:41 +0500
Subject: [PATCH] Paper-Prevent-position-desync-in-playerconnection-causing


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 9f8f33d6f99afb8f93040375a4bac9f21e122551..0a4addf07d7e61267e4292b79f52ba34985d007b 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -657,6 +657,11 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                             }
 
                             this.player.move(EnumMoveType.PLAYER, d7, d8, d9);
+                            // Paper start - prevent position desync
+                            if (this.teleportPos != null) {
+                                return; // ... thanks Mojang for letting move calls teleport across dimensions.
+                            }
+                            // Paper end - prevent position desync
                             this.player.onGround = packetplayinflying.a();
                             double d12 = d8;
 
