From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Tue, 12 Jul 2022 02:02:55 +0500
Subject: [PATCH] Prevent-Village-Doors-from-Loading-Chunks


diff --git a/src/main/java/net/minecraft/server/Village.java b/src/main/java/net/minecraft/server/Village.java
index 6536e5fb843fd6e9f28ad975fc73678fafffff47..6d32ca025d3a2cae463bfac72c0ba73a1644d5a8 100644
--- a/src/main/java/net/minecraft/server/Village.java
+++ b/src/main/java/net/minecraft/server/Village.java
@@ -39,6 +39,31 @@ public class Village {
         this.a = world;
     }
 
+    // NeonPaper start
+    private BlockPosition[] positions = null;
+
+    private void calculateNewCheckPositions() {
+        if(this.d == null || this.d.equals(BlockPosition.ZERO)) {
+            this.positions = null;
+        } else {
+            this.positions = new BlockPosition[] { this.d.a(-this.e, 0, -this.e),
+                    this.d.a(-this.e, 0, this.e),
+                    this.d.a(this.e, 0, -this.e),
+                    this.d.a(this.e, 0, this.e),
+                    this.d};
+        }
+    }
+
+    public boolean isVillageAreaLoaded() {
+        for(int i = 0; this.positions != null && i < this.positions.length; i++) {
+            if(this.a.isLoaded(this.positions[i])) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // NeonPaper end
+
     public void a(World world) {
         this.a = world;
     }
@@ -50,6 +75,7 @@ public class Village {
             return;
         }
         // Paper end
+        if(!this.isVillageAreaLoaded()) { return; } // NeonPaper
         this.g = i;
         this.m();
         this.l();
@@ -368,6 +394,7 @@ public class Village {
 
             this.e = Math.max(32, (int) Math.sqrt((double) j) + 1);
         }
+        this.calculateNewCheckPositions(); // NeonPaper
     }
 
     public int a(String s) {
@@ -422,7 +449,7 @@ public class Village {
                 this.j.put(nbttagcompound2.getString("Name"), Integer.valueOf(nbttagcompound2.getInt("S")));
             }
         }
-
+        this.calculateNewCheckPositions(); // NeonPaper
     }
 
     public void b(NBTTagCompound nbttagcompound) {
