From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Sun, 17 Jul 2022 14:11:01 +0500
Subject: [PATCH] Paper-Backport-Dont-resent-entity-on-art-update


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
index 3e9448a923e7e3c386a9febe3b2a5a6cfc818d29..7b1926e743ee2480ff77b73306aef6eee93cb89d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
@@ -38,7 +38,7 @@ public class CraftPainting extends CraftHanging implements Painting {
             painting.setDirection(painting.direction);
             return false;
         }
-        this.update();
+        //this.update(); Paper - Don't resent entity on art update
         return true;
     }
 
