From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Sun, 17 Jul 2022 14:15:35 +0500
Subject: [PATCH] Paper-Throw-exception-on-world-create-while-being-ticked


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index bf0301a50294bfbcbcc9ce94fc2761dab0981334..2b82bace2a06e2eb22bebe1de92732cfeea05118 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -990,6 +990,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
 
             // this.i[i][this.ticks % 100] = System.nanoTime() - j; // CraftBukkit
         }
+        this.isIteratingOverLevels = false; // Paper
 
         this.methodProfiler.c("connection");
         MinecraftTimings.connectionTimer.startTiming(); // Spigot
@@ -1233,6 +1234,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
     }
 
     public List<String> tabCompleteCommand(ICommandListener icommandlistener, String s, @Nullable BlockPosition blockposition, boolean flag) {
+        this.isIteratingOverLevels = true; // Paper
         /* CraftBukkit start - Allow tab-completion of Bukkit commands
         ArrayList arraylist = Lists.newArrayList();
         boolean flag1 = s.startsWith("/");
@@ -1284,6 +1286,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
     public boolean M() {
         return true; // CraftBukkit
     }
+    public boolean isIteratingOverLevels = false; // Paper
 
     public String getName() {
         return "Server";
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c26b7fc09cd8bbb0b6ba01d46bc9b51e31141540..8dcdc8216098bc6958a1e276dfb4fc564e6f9bc6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -931,6 +931,7 @@ public final class CraftServer implements Server {
 
     @Override
     public World createWorld(WorldCreator creator) {
+        Preconditions.checkState(!this.console.isIteratingOverLevels, "Cannot create a world while worlds are being ticked"); // Paper
         Validate.notNull(creator, "Creator may not be null");
 
         String name = creator.name();
@@ -1049,6 +1050,7 @@ public final class CraftServer implements Server {
 
     @Override
     public boolean unloadWorld(World world, boolean save) {
+        Preconditions.checkState(!this.console.isIteratingOverLevels, "Cannot unload a world while worlds are being ticked"); // Paper
         if (world == null) {
             return false;
         }
