From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Wed, 20 Jul 2022 08:56:58 +0500
Subject: [PATCH] Cache-chunk-coordinations


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index e4838013992cd292e112f000c292348b5ecd4279..69e78e1c0cbc2a4c7ad7ebf206e9e4063e9a2fb2 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -126,6 +126,7 @@ public class Chunk {
         return false;
     }
     // Airplane end
+    private ChunkCoordIntPair chunkCoords; // Dionysus
 
     public Chunk(World world, int i, int j) {
         this.sections = new ChunkSection[16];
@@ -152,6 +153,7 @@ public class Chunk {
         this.chunkKey = ChunkCoordIntPair.a(this.locX, this.locZ);
 
         this.lightningTick = this.world.random.nextInt(100000) << 1; // Airplane - initialize lightning tick
+        this.chunkCoords = new ChunkCoordIntPair(this.locX, this.locZ); // Dionysus
     }
 
     public org.bukkit.Chunk bukkitChunk;
@@ -1238,7 +1240,7 @@ public class Chunk {
     }
 
     public ChunkCoordIntPair k() {
-        return new ChunkCoordIntPair(this.locX, this.locZ);
+        return this.chunkCoords; // Dionysus
     }
 
     public boolean c(int i, int j) {
