From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlexProgrammerDE <40795980+AlexProgrammerDE@users.noreply.github.com>
Date: Sat, 30 Jul 2022 20:38:56 +0200
Subject: [PATCH] Improve position iterator performance


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index a966a83549cf661339778c6ad49bdd46ee5ea64c..99c331d48f98fb38379395b134a63cc9f9e56940 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -1206,9 +1206,11 @@ public class Chunk {
             this.o();
         }
 
-        while (!this.y.isEmpty()) {
+        while (true) { // Dionysus - Improve position iterator performance
             BlockPosition blockposition = (BlockPosition) this.y.poll();
 
+            if (blockposition == null) break; // Dionysus - Improve position iterator performance
+
             if (this.a(blockposition, Chunk.EnumTileEntityState.CHECK) == null && this.getBlockData(blockposition).getBlock().isTileEntity()) {
                 TileEntity tileentity = this.g(blockposition);
 
