From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlexProgrammerDE <40795980+AlexProgrammerDE@users.noreply.github.com>
Date: Sat, 30 Jul 2022 20:38:56 +0200
Subject: [PATCH] Improve position iterator performance


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 4cebe154334214b9a831ffec597f47148b021013..5ed789cd30c8b7d40c2a81863b7a91a62b1704f1 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -1206,9 +1206,11 @@ public class Chunk {
             this.o();
         }
 
-        while (!this.y.isEmpty()) {
+        while (true) { // Dionysus - Improve position iterator performance
             BlockPosition blockposition = (BlockPosition) this.y.poll();
 
+            if (blockposition == null) break; // Dionysus - Improve position iterator performance
+
             if (this.a(blockposition, Chunk.EnumTileEntityState.CHECK) == null && this.getBlockData(blockposition).getBlock().isTileEntity()) {
                 TileEntity tileentity = this.g(blockposition);
 
