From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: comendantmc <2b2t.org.ru@gmail.com>
Date: Thu, 4 Aug 2022 13:10:53 +0200
Subject: [PATCH] Add option to disable communication with mojang servers in
 offline mode


diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperMinecraftSessionService.java b/src/main/java/com/destroystokyo/paper/profile/PaperMinecraftSessionService.java
index 61cfdf73c8a5b0dcf2f9903ebbb2f4132ba1f0dd..8f20768564a972df7b30282684412d8528cdaf35 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperMinecraftSessionService.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperMinecraftSessionService.java
@@ -6,6 +6,9 @@ import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.minecraft.MinecraftProfileTexture;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
 import com.mojang.authlib.yggdrasil.YggdrasilMinecraftSessionService;
+import net.minecraft.server.MinecraftServer;
+import java.util.UUID;
+import com.google.common.base.Charsets;
 
 import java.util.Map;
 
@@ -34,6 +37,9 @@ public class PaperMinecraftSessionService extends YggdrasilMinecraftSessionServi
 
     @Override
     protected GameProfile fillGameProfile(GameProfile profile, boolean requireSecure) {
-        return super.fillGameProfile(profile, requireSecure);
+        if (!dev.pomf.dionysus.DionysusConfig.disableMojangServerCommunicationInOfflineMode || MinecraftServer.getServer().getOnlineMode())
+            return super.fillGameProfile(profile, requireSecure);
+
+        return profile;
     }
 }
diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index fc27044e8643afb828a71fd7178f865760d96528..86d3ff8e04b9e410d6c3de09786f685d2a1e5412 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -397,4 +397,9 @@ public class DionysusConfig {
                 "and \"Deleting duplicate entity\" log messages."
         );
     }
+
+    public static boolean disableMojangServerCommunicationInOfflineMode = false;
+    private static void disableMojangServerCommunicationInOfflineMode() {
+        disableMojangServerCommunicationInOfflineMode = getBoolean("disable-mojang-server-communication-in-offline-mode", false);
+    }
 }
