From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlexProgrammerDE <40795980+AlexProgrammerDE@users.noreply.github.com>
Date: Sun, 21 Aug 2022 18:53:15 +0200
Subject: [PATCH] Add more information to crash report


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftCrashReport.java b/src/main/java/org/bukkit/craftbukkit/CraftCrashReport.java
index 67f285668cd29108c18c059339c8388d6d274ea8..a15148259aaabfa80298e660d687b4b90010b138 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftCrashReport.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftCrashReport.java
@@ -5,9 +5,12 @@ import java.io.StringWriter;
 import java.util.Arrays;
 import java.util.Map;
 import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
 import net.minecraft.server.CrashReportCallable;
 
 import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.scheduler.CraftTask;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.PluginDescriptionFile;
 
@@ -31,6 +34,8 @@ public class CraftCrashReport implements CrashReportCallable<Object> {
                 value.append(' ').append(entry.getKey().getState().name()).append(' ').append(entry.getKey().getName()).append(": ").append(Arrays.toString(entry.getValue())).append(',');
             }
             value.append("}\n   ").append(Bukkit.getScheduler().toString());
+            value.append("Pending tasks: ").append(Bukkit.getScheduler().getPendingTasks().stream().map(task -> "{" + task.getTaskId() + " " + task.getOwner().getName() + "}").collect(Collectors.joining(", ")));
+            value.append("Active workers: ").append(Bukkit.getScheduler().getActiveWorkers().stream().map(task -> "{" + task.getTaskId() + " " + task.getOwner().getName() + "}").collect(Collectors.joining(", ")));
         } catch (Throwable t) {
             value.append("\n   Failed to handle CraftCrashReport:\n");
             PrintWriter writer = new PrintWriter(value);
