From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RemainingToast <Remainingtoast@gmail.com>
Date: Thu, 25 Aug 2022 00:20:57 +1000
Subject: [PATCH] save books as json


diff --git a/pom.xml b/pom.xml
index 1f2fc8223808ce352267660ca9a11d7ffec94d70..362b4caa6bf3cbf4eee041a91d46748f06e4b799 100644
--- a/pom.xml
+++ b/pom.xml
@@ -75,6 +75,15 @@
             <version>1.3.0</version>
         </dependency>
 
+        <!-- Dionysus start - Save Books as Json -->
+        <dependency>
+            <groupId>de.tr7zw</groupId>
+            <artifactId>item-nbt-api</artifactId>
+            <version>2.10.0-SNAPSHOT</version>
+            <scope>compile</scope>
+        </dependency>
+        <!-- Dionysus end -->
+
         <!--
           Required to add the missing Log4j2Plugins.dat file from log4j-core
           which has been removed by Mojang. Without it, log4j has to classload
@@ -303,6 +312,12 @@
                             </filters>
                             <!-- Paper end - Update Log4j -->
                             <relocations>
+                                <!-- Dionysus start - save books as json -->
+                                <relocation>
+                                    <pattern>de.tr7zw.changeme.nbtapi</pattern>
+                                    <shadedPattern>dev.pomf.dionysus.nbtapi</shadedPattern>
+                                </relocation>
+                                <!-- Dionysus end -->
                                 <!-- Paper - Workaround for hardcoded path lookup in dependency, easier than forking it - GH-189 -->
                                 <!--<relocation>-->
                                     <!--<pattern>joptsimple</pattern>-->
diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 48319f349c63bf9d04197a48ef756b8c7fdd17a3..45b8bb37fb94b3a1993654d80a5e36ad1dc4f245 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -444,4 +444,13 @@ public class DionysusConfig {
             networks is predictable and intuitive rather than locational and chaotic."""
         );
     }
+
+    public static boolean saveBooksAsJson = true;
+    private static void saveBooksAsJson() {
+        saveBooksAsJson = getBoolean("save-books-as-json", saveBooksAsJson);
+        setComment(
+                "save-books-as-json",
+                "Books stored in /world/bookdata/*"
+        );
+    }
 }
diff --git a/src/main/java/dev/pomf/dionysus/util/BookUtil.java b/src/main/java/dev/pomf/dionysus/util/BookUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..2581aac0d184403df2bffeca610d75b10af09ae6
--- /dev/null
+++ b/src/main/java/dev/pomf/dionysus/util/BookUtil.java
@@ -0,0 +1,156 @@
+package dev.pomf.dionysus.util;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import de.tr7zw.changeme.nbtapi.NBTItem;
+import org.bukkit.Material;
+import org.bukkit.block.ShulkerBox;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.BlockStateMeta;
+import org.bukkit.inventory.meta.BookMeta;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.Writer;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BookUtil {
+
+    private static final String FILE_PATH = System.getProperty("user.dir") + "/world/bookdata/%s.json";
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+
+    public static void loadInventory(Player player, Inventory inventory) {
+        for (ItemStack slot : inventory.getContents()) {
+            if (slot == null) continue;
+            loadItemStack(slot);
+        }
+
+        player.updateInventory();
+    }
+
+    public static void unloadInventory(Player player, Inventory inventory) {
+        for (ItemStack slot : inventory.getContents()) {
+            if (slot == null) continue;
+            unloadItemStack(slot);
+            if (slot.getItemMeta() instanceof BlockStateMeta meta) {
+                if (meta instanceof ShulkerBox) {
+                    ShulkerBox shulker = (ShulkerBox) meta.getBlockState();
+                    Inventory shulkerInv = shulker.getInventory();
+                    for (ItemStack shulkerSlot : shulkerInv.getContents()) {
+                        if (shulkerSlot == null) continue;
+                        unloadItemStack(shulkerSlot);
+                    }
+                    shulker.getInventory().setContents(shulkerInv.getContents());
+                    meta.setBlockState(shulker);
+                    slot.setItemMeta(meta);
+                }
+            }
+        }
+
+        player.updateInventory();
+    }
+
+    public static void unloadItemStack(ItemStack itemStack) {
+        if (isBook(itemStack)) {
+            NBTItem nbtItem = new NBTItem(itemStack);
+            if (!(nbtItem.hasKey("key") && nbtItem.getBoolean("unloaded"))) {
+                itemStack.setItemMeta(unloadBook(itemStack).getItemMeta());
+            }
+        }
+    }
+
+    public static void loadItemStack(ItemStack itemStack) {
+        if (isBook(itemStack)) {
+            NBTItem nbtItem = new NBTItem(itemStack);
+            if (nbtItem.getBoolean("unloaded")) {
+                itemStack.setItemMeta(loadBook(itemStack).getItemMeta());
+            }
+        }
+    }
+
+    private static boolean isBook(ItemStack item) {
+        return item.getType() == Material.BOOK_AND_QUILL || item.getType() == Material.WRITTEN_BOOK;
+    }
+
+    private static ItemStack unloadBook(ItemStack book) {
+        ItemStack clone = book;
+        BookMeta meta = (BookMeta) book.getItemMeta();
+        try {
+            List<String> pages = new ArrayList<>();
+            for(String str: meta.getPages()) {
+                ByteBuffer buffer = StandardCharsets.UTF_8.encode(str);
+                String utf8str = StandardCharsets.UTF_8.decode(buffer).toString();
+                pages.add(utf8str);
+            }
+            Path path = Paths.get(FILE_PATH.formatted(hash(meta.getPages()+"")));
+            if (!Files.exists(path.getParent())) {
+                Files.createDirectory(path.getParent());
+            }
+            Writer writer = Files.newBufferedWriter(path);
+            GSON.toJson(new Book(pages), writer);
+            writer.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        NBTItem nbtItem = new NBTItem(clone);
+        nbtItem.setString("key", hash(meta.getPages()+""));
+        nbtItem.setBoolean("unloaded", true);
+        clone = nbtItem.getItem();
+        meta = (BookMeta) clone.getItemMeta();
+        meta.setPages(new ArrayList<>());
+        clone.setItemMeta(meta);
+        return clone;
+    }
+
+    private static ItemStack loadBook(ItemStack book) {
+        ItemStack clone = book;
+        NBTItem nbtItem = new NBTItem(clone);
+        try {
+            Reader reader = Files.newBufferedReader(Paths.get(FILE_PATH.formatted(nbtItem.getString("key"))));
+            Book bookObject = GSON.fromJson(reader, Book.class);
+            reader.close();
+            BookMeta bookMeta = (BookMeta) clone.getItemMeta();
+            bookMeta.setPages(bookObject.getPages());
+            clone.setItemMeta(bookMeta);
+            nbtItem = new NBTItem(clone);
+            nbtItem.setBoolean("unloaded", false);
+            clone = nbtItem.getItem();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return clone;
+    }
+
+    private static String hash(String md5) {
+        try {
+            StringBuilder hash = new StringBuilder();
+            for (byte byte_ : java.security.MessageDigest.getInstance("MD5").digest(md5.getBytes())) {
+                hash.append(Integer.toHexString((byte_ & 0xFF) | 0x100), 1, 3);
+            }
+            return hash.toString();
+        } catch (java.security.NoSuchAlgorithmException ignored) {}
+        return null;
+    }
+
+    public static class Book {
+
+        private final List<String> pages;
+
+        public Book(List<String> pages) {
+            this.pages = pages;
+        }
+
+        public List<String> getPages() {
+            return pages;
+        }
+    }
+
+}
diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 4fafb4977a40176d88e790924afba8fa9e17867b..1029907ddd4fcfd336e548e0d9587ad814246516 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -2,9 +2,13 @@ package net.minecraft.server;
 
 import java.util.Iterator;
 import javax.annotation.Nullable;
+
+import dev.pomf.dionysus.DionysusConfig;
+import dev.pomf.dionysus.util.BookUtil;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 // CraftBukkit start
+import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityPickupItemEvent;
 import org.bukkit.event.player.PlayerPickupItemEvent;
 // CraftBukkit end
@@ -374,6 +378,12 @@ public class EntityItem extends Entity implements HopperPusher {
                     return;
                 }
 
+                // Dionysus start
+                if (DionysusConfig.saveBooksAsJson) {
+                    BookUtil.loadInventory((Player) entityhuman.getBukkitEntity(), entityhuman.getBukkitEntity().getInventory());
+                }
+                // Dionysus end
+
                 // Call newer event afterwards
                 EntityPickupItemEvent entityEvent = new EntityPickupItemEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
                 entityEvent.setCancelled(!entityhuman.canPickUpLoot);
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 099b3ebb6a8be90e0cbf45b4a21be6f918153a54..ca0672ce8cb77e1fb67a3f820bd96a678c2afb84 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -2,6 +2,8 @@ package net.minecraft.server;
 
 import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
+import dev.pomf.dionysus.DionysusConfig;
+import dev.pomf.dionysus.util.BookUtil;
 import io.netty.buffer.Unpooled;
 import java.util.ArrayDeque; // Paper
 import java.util.ArrayList;
@@ -22,6 +24,7 @@ import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.Player;
 import org.bukkit.event.inventory.InventoryType;
 import org.bukkit.event.player.PlayerChangedMainHandEvent;
 import org.bukkit.event.player.PlayerGameModeChangeEvent;
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index a474d8af1670ca1619bc0043f2b4f912ea9a4844..66373fe463cb70c75a536cbc38092fa7760eab88 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -2,6 +2,11 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import java.util.ArrayList;
+
+import dev.pomf.dionysus.DionysusConfig;
+import dev.pomf.dionysus.util.BookUtil;
+import org.bukkit.block.BlockState;
+import org.bukkit.block.ShulkerBox;
 import org.bukkit.event.block.BlockBreakEvent;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.Event;
@@ -307,6 +312,15 @@ public class PlayerInteractManager {
                 ((EntityPlayer) this.player).playerConnection.sendPacket(packet);
             }
 
+            // Dionysus start - save books as jsons
+            if (DionysusConfig.saveBooksAsJson) {
+                BlockState state = block.getState();
+                if (state instanceof ShulkerBox shulker) {
+                    BookUtil.unloadInventory(player.getBukkitEntity(), shulker.getInventory());
+                }
+            }
+            // Dionysus end
+
             event = new BlockBreakEvent(block, this.player.getBukkitEntity());
 
             // Sword + Creative mode pre-cancel
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 897a49eda09df4654ea31808907cb60c8fa94f77..082d17e2fa69c669f076b37f809d272af57c4f55 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -6,6 +6,7 @@ import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
 import com.mojang.authlib.GameProfile;
 import dev.pomf.dionysus.DionysusConfig;
+import dev.pomf.dionysus.util.BookUtil;
 import io.netty.buffer.Unpooled;
 import java.io.File;
 import java.net.SocketAddress;
@@ -32,8 +33,10 @@ import org.bukkit.craftbukkit.chunkio.ChunkIOExecutor;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.TravelAgent;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.craftbukkit.util.CraftChatMessage;
 import org.bukkit.entity.Player;
+import org.bukkit.event.block.BlockBreakEvent;
 import org.bukkit.event.player.PlayerChangedWorldEvent;
 import org.bukkit.event.player.PlayerPortalEvent;
 import org.bukkit.event.player.PlayerJoinEvent;
@@ -41,6 +44,7 @@ import org.bukkit.event.player.PlayerLoginEvent;
 import org.bukkit.event.player.PlayerQuitEvent;
 import org.bukkit.event.player.PlayerRespawnEvent;
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
+import org.bukkit.inventory.PlayerInventory;
 import org.bukkit.util.Vector;
 import org.spigotmc.event.player.PlayerSpawnLocationEvent;
 // CraftBukkit end
@@ -435,6 +439,20 @@ public abstract class PlayerList {
         // this.sendAll(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER, new EntityPlayer[] { entityplayer})); // CraftBukkit - replaced with loop below
         WorldServer worldserver = this.server.getWorldServer(entityplayer.dimension);
 
+        // Dionysus start - save books as json
+        if (DionysusConfig.saveBooksAsJson) {
+            final CraftPlayer p = entityplayer.getBukkitEntity();
+            final PlayerInventory i = p.getInventory();
+
+            // Unload all books on join
+            BookUtil.unloadInventory(p, i);
+
+            // Load books in hands
+            BookUtil.loadItemStack(i.getItemInMainHand());
+            BookUtil.loadItemStack(i.getItemInOffHand());
+        }
+        // Dionysus end
+
         // CraftBukkit start
         PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(cserver.getPlayer(entityplayer), joinMessage);
         cserver.getPluginManager().callEvent(playerJoinEvent);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f1a3ca9509ebc50eaa67bf4dac2645f5ee37bbef..78f90acbec73130309c3fb6d6d201f8f992afffb 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -10,6 +10,9 @@ import javax.annotation.Nullable;
 import com.google.common.base.Function;
 import com.google.common.base.Functions;
 
+import de.tr7zw.changeme.nbtapi.NBTItem;
+import dev.pomf.dionysus.DionysusConfig;
+import dev.pomf.dionysus.util.BookUtil;
 import net.minecraft.server.*;
 
 import org.bukkit.Bukkit;
@@ -19,6 +22,7 @@ import org.bukkit.Statistic.Type;
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockFace;
 import org.bukkit.block.BlockState;
+import org.bukkit.block.ShulkerBox;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.craftbukkit.CraftStatistic;
 import org.bukkit.craftbukkit.CraftWorld;
@@ -30,6 +34,7 @@ import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.craftbukkit.inventory.CraftMetaBook;
+import org.bukkit.craftbukkit.scheduler.CraftScheduler;
 import org.bukkit.craftbukkit.util.CraftDamageSource;
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.entity.AreaEffectCloud;
@@ -147,6 +152,15 @@ public class CraftEventFactory {
             equipmentSlot = EquipmentSlot.OFF_HAND;
         }
 
+        // Dionysus start - save books as jsons
+        if (DionysusConfig.saveBooksAsJson) {
+            BlockState state = placedBlock.getState();
+            if (state instanceof ShulkerBox shulker) {
+                BookUtil.unloadInventory(player, shulker.getInventory());
+            }
+        }
+        // Dionysus end
+
         BlockPlaceEvent event = new BlockPlaceEvent(placedBlock, replacedBlockState, blockClicked, item, player, canBuild, equipmentSlot);
         craftServer.getPluginManager().callEvent(event);
 
@@ -820,6 +834,14 @@ public class CraftEventFactory {
         CraftPlayer craftPlayer = player.getBukkitEntity();
         player.activeContainer.transferTo(container, craftPlayer);
 
+        // Dionysus start - save books as json
+        if (DionysusConfig.saveBooksAsJson) {
+            InventoryView inv = player.activeContainer.getBukkitView();
+            BookUtil.unloadInventory(player.getBukkitEntity(), inv.getTopInventory());
+            BookUtil.unloadInventory(player.getBukkitEntity(), craftPlayer.getInventory());
+        }
+        // Dionysus end
+
         InventoryOpenEvent event = new InventoryOpenEvent(container.getBukkitView());
         event.setCancelled(cancelled);
         server.getPluginManager().callEvent(event);
@@ -977,6 +999,11 @@ public class CraftEventFactory {
         handleInventoryCloseEvent(human, org.bukkit.event.inventory.InventoryCloseEvent.Reason.UNKNOWN);
     }
     public static void handleInventoryCloseEvent(EntityHuman human, org.bukkit.event.inventory.InventoryCloseEvent.Reason reason) {
+        // Dionysus start - save books as json
+        if (DionysusConfig.saveBooksAsJson) {
+            BookUtil.loadInventory((Player) human.getBukkitEntity(), human.getBukkitEntity().getInventory());
+        }
+        // Dionysus end
         InventoryCloseEvent event = new InventoryCloseEvent(human.activeContainer.getBukkitView(), reason);
         // Paper end
         human.world.getServer().getPluginManager().callEvent(event);
