From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Sun, 16 Oct 2022 14:07:41 +0500
Subject: [PATCH] fix-setViewDistance-bug


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 9d836e7e12284c1c2bf72f4cc5b4c17802425835..091a83f8cdea83231d27892f8a0546dd2bccf6fe 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -431,9 +431,11 @@ public class PlayerChunkMap {
         int oldViewDistance = entityplayer.getViewDistance();
         if (i != oldViewDistance) {
             int j = i - oldViewDistance;
-            
-            int k = (int) entityplayer.locX >> 4;
-            int l = (int) entityplayer.locZ >> 4;
+
+            // Reaper start - Calculate by last managedPos, not by current location, fixes getting stuck in the PlayerChunk viewers list
+            int k = (int) entityplayer.d >> 4;
+            int l = (int) entityplayer.e >> 4;
+            // Reaper end
             int i1;
             int j1;
 
@@ -451,7 +453,7 @@ public class PlayerChunkMap {
                 for (i1 = k - oldViewDistance; i1 <= k + oldViewDistance; ++i1) {
                     for (j1 = l - oldViewDistance; j1 <= l + oldViewDistance; ++j1) {
                         if (!this.a(i1, j1, k, l, i)) {
-                            this.c(i1, j1).b(entityplayer);
+                            PlayerChunk chunk = this.getChunk(i1, j1); if (chunk != null) chunk.b(entityplayer); // Reaper - Fix memory leak
                         }
                     }
                 }
