From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruviolence@users.noreply.github.com>
Date: Sun, 16 Oct 2022 14:31:53 +0500
Subject: [PATCH] fix-map-id-overflow

Co-authored-by: RemainingToast <remainingtoast@gmail.com>

diff --git a/src/main/java/net/minecraft/server/PersistentCollection.java b/src/main/java/net/minecraft/server/PersistentCollection.java
index 50056f49a8eaea687d00175e41a877040934e9ae..697a6687e51745f79791a1514ea5be02fc726e42 100644
--- a/src/main/java/net/minecraft/server/PersistentCollection.java
+++ b/src/main/java/net/minecraft/server/PersistentCollection.java
@@ -151,6 +151,7 @@ public class PersistentCollection {
             oshort = Short.valueOf((short) 0);
         } else {
             oshort = Short.valueOf((short) (oshort.shortValue() + 1));
+            oshort = (short) Math.min(0, oshort); // Reaper - Fix map ID overflow
         }
 
         this.d.put(s, oshort);
