From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruviolence@users.noreply.github.com>
Date: Sun, 16 Oct 2022 14:31:53 +0500
Subject: [PATCH] [Reaper 0131] Fix map id overflow

Co-authored-by: RemainingToast <remainingtoast@gmail.com>

diff --git a/src/main/java/net/minecraft/server/PersistentCollection.java b/src/main/java/net/minecraft/server/PersistentCollection.java
index 50056f49a8eaea687d00175e41a877040934e9ae..5cebee7d5ec0bf8b710e11776944327cd88aa698 100644
--- a/src/main/java/net/minecraft/server/PersistentCollection.java
+++ b/src/main/java/net/minecraft/server/PersistentCollection.java
@@ -133,7 +133,7 @@ public class PersistentCollection {
                     if (nbtbase instanceof NBTTagShort) {
                         NBTTagShort nbttagshort = (NBTTagShort) nbtbase;
                         short short0 = nbttagshort.f();
-
+                        short0 = (short) Math.min(0, Math.abs(short0)); // Dionysus - Fix map ID overflow
                         this.d.put(s, Short.valueOf(short0));
                     }
                 }
@@ -151,6 +151,7 @@ public class PersistentCollection {
             oshort = Short.valueOf((short) 0);
         } else {
             oshort = Short.valueOf((short) (oshort.shortValue() + 1));
+            oshort = (short) Math.min(0, Math.abs(oshort)); // Reaper - Fix map ID overflow
         }
 
         this.d.put(s, oshort);
