From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoftikLord <dimap9986@gmail.com>
Date: Wed, 23 Nov 2022 21:27:25 +0500
Subject: [PATCH] feat-max-view-distance-from-32-to-64


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 091a83f8cdea83231d27892f8a0546dd2bccf6fe..c34994c018f964ad2d02747fe39f194ad8bd64ab 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -405,7 +405,7 @@ public class PlayerChunkMap {
     public final void setViewDistanceForAll(int viewDistance) { this.a(viewDistance); } // Paper - OBFHELPER
     // Paper start - Separate into two methods
     public void a(int i) {
-        i = MathHelper.clamp(i, 3, 32);
+        i = MathHelper.clamp(i, 1, 64);
         if (i != this.j) {
             int j = i - this.j;
             ArrayList arraylist = Lists.newArrayList(this.managedPlayers);
@@ -427,7 +427,7 @@ public class PlayerChunkMap {
     
     // Copied from above with minor changes
     public void setViewDistance(EntityPlayer entityplayer, int i, boolean markSort) {
-        i = MathHelper.clamp(i, 3, 32);
+        i = MathHelper.clamp(i, 1, 64);
         int oldViewDistance = entityplayer.getViewDistance();
         if (i != oldViewDistance) {
             int j = i - oldViewDistance;
@@ -557,7 +557,7 @@ public class PlayerChunkMap {
 
         // This represents the view distance that we will set on the player
         // It can exist as a negative value
-        int playerViewDistance = MathHelper.clamp(distanceIn, 3, 32);
+        int playerViewDistance = MathHelper.clamp(distanceIn, 1, 64);
 
         // This value is the one we actually use to update the chunk map
         // We don't ever want this to be a negative
